clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -224,11 +224,16 @@ public void onFailure(RuntimeException e) {
      * @param partitions the partitions to reset
      */
     public void resetOffsetsIfNeeded(Set<TopicPartition> partitions) {
+        List<TopicPartition> failedPartitions = new ArrayList<>();
         for (TopicPartition tp : partitions) {
             // TODO: If there are several offsets to reset, we could submit offset requests in parallel
-            if (subscriptions.isAssigned(tp) && subscriptions.isOffsetResetNeeded(tp))
-                resetOffset(tp);
+            if (subscriptions.isAssigned(tp) && subscriptions.isOffsetResetNeeded(tp)) {
+                if (!resetOffset(tp)) {
+                    failedPartitions.add(tp);
+                }
+            }
```

Comment:
I think this `if` block can be written this way too so we can avoid the extra nesting (since Java performs short-circuit evaluation for `&&` and `||`):
```
        if (subscriptions.isAssigned(tp) && subscriptions.isOffsetResetNeeded(tp) && !resetOffset(tp))
            failedPartitions.add(tp);
```